Add std::thread::available_concurrency

yoshuawuyts opened the pull request:
This PR adds a counterpart to [C++'s `std::thread::hardware_concurrency`](https://en.cppreference.com/w/cpp/thread/thread/hardware_concurrency) to Rust, tracking issue https://github.com/rust-lang/rust/issues/74479.

cc/ @rust-lang/libs 

## Motivation

Being able to know how many hardware threads a platform supports is a core part of building multi-threaded code. In C++ 11 this has become available through the [`std::thread::hardware_concurrency`](https://en.cppreference.com/w/cpp/thread/thread/hardware_concurrency) API. Currently in Rust most of the ecosystem depends on the [`num_cpus` crate](https://docs.rs/num_cpus/1.13.0/num_cpus/) ([no.35 in top 500 crates](https://docs.google.com/spreadsheets/d/1wwahRMHG3buvnfHjmPQFU4Kyfq15oTwbfsuZpwHUKc4/edit#gid=1253069234)) to provide this functionality. This PR proposes an API to provide access to the number of hardware threads available on a given platform.

__edit (2020-07-24):__ The purpose of this PR is to provide a hint for how many threads to spawn to saturate the processor. There's value in introducing APIs for NUMA and Windows processor groups, but those are intentionally out of scope for this PR. See: https://github.com/rust-lang/rust/pull/74480#issuecomment-662116186.

## Naming

Discussing the naming of the API on Zulip surfaced two options:

- `std::thread::hardware_concurrency`
- `std::thread::hardware_threads`

Both options seemed acceptable, but overall people seem to gravitate the most towards `hardware_threads`. Additionally @jonas-schievink pointed out that the "hardware threads" terminology is well-established and is used in among other the [RISC-V specification](https://riscv.org/specifications/isa-spec-pdf/) (page 20):

> A component is termed a core if it contains an independent instruction fetch unit. A RISC-V-compatible core might support multiple RISC-V-compatible __hardware threads__, or harts, through multithreading.

It's also worth noting that [the original paper introducing C++'s `std::thread` submodule](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2007/n2320.html) unfortunately doesn't feature any discussion on the naming of `hardware_concurrency`, so we can't use that to help inform our decision here.

## Return type

An important consideration @joshtriplett brought up is that we don't want to default to `1` for platforms where the number of available threads cannot be retrieved. Instead we want to inform the users of the fact that we don't know and allow them to handle that case. Which is why this PR uses `Option<NonZeroUsize>` as its return type, where `None` is returned on platforms where we don't know the number of hardware threads available.

The reasoning for `NonZeroUsize` vs `usize` is that if the number of threads for a platform are known, they'll always be at least 1. As evidenced by the example the `NonZero*` family of APIs may currently not be the most ergonomic to use, but improving the ergonomics of them is something that I think we can address separately.

## Implementation

@Mark-Simulacrum pointed out that most of the code we wanted to expose here was already available under `libtest`. So this PR mostly moves the internal code of libtest into a public API.

rust-highfive commented:
r? @dtolnay

(rust_highfive has picked a reviewer for you, use r? to override)

luser commented:
To clarify, would this report the number of "logical" cores on CPUs that support hyperthreading? That's probably the right choice if the API intends to answer the question "how much parallelism is reasonable".

yoshuawuyts commented:
> To clarify, would this report the number of "logical" cores on CPUs that support hyperthreading? 

@luser Yes I believe it should. Since I didn't write the backing impl I had to verify, but the `num_cpus::get` implementation takes an identical approach as the code in rustc (as opposed to the `num_cpus::get_physical`).

Perhaps we should clarify in the docs what we mean by "hardware threads"?

dfamonteiro commented:
> To clarify, would this report the number of "logical" cores on CPUs that support hyperthreading? That's probably the right choice if the API intends to answer the question "how much parallelism is reasonable".

That is a great point. Related to this, is it worthwhile to have two functions that distinguish between the number of _logical cores_ and the number of _hardware cores_?

ehuss commented:
Would you mind retaining the original code that used `cfg_if`?  It intentionally was written with a fall-through.

Also, my impression is that all platform-specific code should be in the `sys` module. I'm not sure how much that is enforced, though.

Might also want to address, the `num_cpus` code has been more maintained and supports more than this does. Any particular reason not to use that?

mati865 commented:
Hardware threads is correct term but can be quite confusing for some people. To make it easier you can view it as:
- software threads - created by app inside OS, is not related to amount of CPU threads
- hardware threads - in a great simplification: how many threads CPU can handle simultaneously (for SMT CPUs it's cores*<threads per core>)
- cores - physical cores, SMT threads are not included

guswynn commented:
This doesn't seem to have support for target_os="none", unless I'm totally missing how cfg(target_os) works?

yoshuawuyts commented:
What @mati865 says is accurate. Running the `num_cpus` APIs on my AMD computer yields the following results:

```rust
fn main() {
    println!("hardware threads {}", num_cpus::get());
    println!("hardware cores {}", num_cpus::get_physical());
}
```
```txt
hardware threads: 24
hardware cores: 12
```

@dfamonteiro I'd like to keep the scope of this PR minimal; but it's certainly possible an API along the lines of `std::thread::hardware_cores` could be introduced in the future. One takeaway I'm having from this conversion so far is that we need to do a better job at explaining what a "hardware thread" in the documentation.

yoshuawuyts commented:
@ehuss ah yes, apologies -- seems I missed that in the rebase. Fixed it!

@guswynn it now does (:

vertexclique commented on src/libstd/thread/hardware_threads.rs:
```diff
@@ -0,0 +1,122 @@
+use crate::num::NonZeroUsize;
+
+/// Returns the number of hardware threads available to the process.
```
Logically correct but in practice this should be smth like:
```
Returns the number of hardware threads available on CPU to OS.
```

yoshuawuyts replied:
@vertexclique Can you elaborate? Are you trying to say this is the amount of threads available to the _operating system_ rather than the process?

This would've be accurate for non-OS VMs such as WASI runtimes, and process-based resource restriction for example through `cgroups`.

vertexclique replied:
Even in that scenario, existence of fork call makes more processes. So it might be a single process or not. That is basically ignoring the existence of fork and CommandExt.

Moreover, cgroups are not for single process. You can restrict a resource allocation to a single process or multiple processes. But if we are talking about cgroups then this will not show all exposed hardware threads to process/processes when cpuset is used.

+ One more thing, when thread affinity mask applied you won't get the all cores exposed to the process/processes. Same applies to programs using Numa and deployed with affinity mask applied.

Yet another thing is, after mask applied you will get vcores assigned to you. Not what OS gives then.

In the light of these, writing correct comment for this is important. It is not hardware threads, it is not os exposing and there are some other runtimes which is used as target arch and don't have OS.

Oh btw, check out sched_setaffinity for solving the last two problems.

vertexclique replied:
In addition to that, is this returning nvptx threads offloaded on gpu? Or any thread count(aka gpu lanes) per process/processes on GPUs like openmp does?

yoshuawuyts replied:
@vertexclique Thanks for the explainer! What I've taken away is that there are many moving parts involved in determining how many threads are available to a program. Also it's hard to find a nomer that's both accurate and to the point.

I'm rewording the comment to use the word "program" rather than the more specific "process", and filing a note in the tracking issue that we need to do a pass on terminology before stabilization.

the8472 replied:
This isn't just about cpusets/affinity. cgroups can also apply timeslice quotas. https://www.kernel.org/doc/Documentation/scheduler/sched-bwc.txt
You probably don't want to run more threads than the quota since it would otherwise freeze your process every time the quota is exhausted early. I.e. background work could impact the interactive parts.

the8472 commented:
> What @mati865 says is accurate. Running the `num_cpus` APIs on my AMD computer yields the following results:
> 
> ```rust
> fn main() {
>     println!("hardware threads {}", num_cpus::get());
>     println!("hardware cores {}", num_cpus::get_physical());
> }
> ```
> 
> ```
> hardware threads: 24
> hardware cores: 12
> ```

num_cpus does not actually tell you the number of hardware threads, it tells you the maximum scheduling capacity available to the current process, which can be lower than the number of hardware threads if CPU affinities, cgroups or similar things are applied. This probably is what most thread pools actually want.

Returning the number of hardware threads would lead to oversubscription in containers.

Java does the same: https://www.docker.com/blog/improved-docker-container-integration-with-java-10/

SimonSapin commented:
> `None` is returned on platforms where we don't know the number of hardware threads available.

This `Option::None` value represents not the absence of hardware thread, but a failure to retrieve information about them. It should therefore be `Result::Err(something)` instead.

> we don't want to default to `1` for platforms where the number of available threads cannot be retrieved

For the same reason, it seems undesirable to return `1` on supported platforms when the corresponding `libc` call returns an error. Why not return also return `Err(something)` in those situations? `std::io::Error` might be appropriate.

yoshuawuyts commented:
Updated with @SimonSapin and Leo Le Boueter's feedback; we now return `io::Result<NonZeroUsize>` because syscalls may fail and we should forward the error.

retep998 commented:
There needs to be a loud section about the limitations of this API on Windows. Specifically that it only returns the number of logical processors in the current processor group which is limited to at most 64 logical processors. To support more than 64 logical processors requires explicit support for enumerating processor groups and assigning threads to logical processors in other groups.

https://docs.microsoft.com/en-us/windows/win32/procthread/processor-groups

joshtriplett commented:
@retep998 Rather than having that limitation, we should be calling the APIs that let us figure out the total number of hardware threads, in all processor groups.

yoshuawuyts commented:
@joshtriplett @retep998 Added an entry on the tracking issue under "known issues" that the amount of threads reported on Windows is at most 64. Also tracking @luser's report that the interaction with CPU affinity on Linux doesn't seem right.

mati865 commented:
@joshtriplett and place warning that users will be limited to 64 hardware threads unless they apply Windows specific workaround. I suppose they will rather use crate that does it for them...

retep998 commented:
@joshtriplett Threads will not run on other processor groups by default. Normally threads only have access to the logical processors in the processor group for that process, so `std::thread::hardware_threads` should not count logical processors in other groups. If we want to count logical processors outside the current processor group, then we need to provide a more comprehensive API that reports information on processor groups and NUMA nodes and CPU sets, and have APIs to assign threads appropriately.

joshtriplett commented:
@retep998 Is it possible for a thread's affinity to be for multiple processor groups at once, or does a thread *have* to be limited to a single processor group at a time no matter how it's started or modified? If the former, we could use that and start threads that can run anywhere by default. If the latter, do Rayon and other libraries use the appropriate APIs at the moment to start threads on every hardware thread across the system?

Long-term, I'd like us to have NUMA-aware APIs as well, but for the moment, it'd be nice to have APIs for the simple case of "start as many hardware threads as the system has".

retep998 commented:
@joshtriplett You must explicitly assign the group affinity of a thread using [`SetThreadGroupAffinity`](https://docs.microsoft.com/en-us/windows/win32/api/processtopologyapi/nf-processtopologyapi-setthreadgroupaffinity). A thread can only be assigned to a single group at any given time. [Rayon is *not* multi-group aware, and will only scale up to the 64 logical processors of the processor group the process is started on](https://github.com/rayon-rs/rayon/issues/723). Reporting information on logical processors in other groups should only be done in tandem with allowing users to run their threads on those other processor groups.

joshtriplett commented:
@retep998 That's an unfortunate limitation, and it sounds like there's no way for us to do better automatically. In that case, yeah, we should go ahead and list the number of processors in the processor group, but we should also try to offer a simple way to spawn threads on every CPU on the system.

SimonSapin commented:
A typical use of this function is to decide how many software threads to spawn for some parallel computation. Therefore what matters is not how much hardware exists somewhere, but how much of it can actually be used. Windows above 64 threads is an example where these two things are not the same.

Because of this, I feel that the function’s name should *not* include the word "hardware". Something like `available_concurrency` sounds more accurate.

joshtriplett commented:
On Tue, Jul 21, 2020 at 11:56:27AM -0700, Simon Sapin wrote:
> A typical use of this function is to decide how many software threads to spawn for some parallel computation. Therefore what matters is not how much hardware exists somewhere, but how much of it can actually be used. Windows above 64 threads is an example where these two things are not the same.
> 
> Because of this, I feel that the function’s name should *not* include the word "hardware". Something like `available_concurrency` sounds more accurate.

That seems reasonable. Another case where these might differ would be
when running in a Linux cgroup that limits available CPUs.

Amanieu commented:
I'm increasingly getting the feeling that this is a hard problem and that we will likely not end up with the best API (as seen with hindsight 2 years from now).

This is the kind of API that is best left to crates.io where breaking changes can be made as needed by simply bumping the version number. On the other hand the standard library API must remain stable for *ever* which is unfortunate if we want to make a breaking change.

joshtriplett commented:
On Tue, Jul 21, 2020 at 02:15:27PM -0700, Amanieu wrote:
> I'm increasingly getting the feeling that this is a hard problem and that we will likely not end up with the best API (as seen with hindsight 2 years from now).
> 
> This is the kind of API that is best left to crates.io where breaking changes can be made as needed by simply bumping the version number. On the other hand the standard library API must remain stable for *ever* which is unfortunate if we want to make a breaking change.

I think there's value in having a simple API that's either "return a
number or an error", and having that API match up with "create that many
threads" using the standard library's thread creation mechanisms. That
interface seems simple and stable enough to go in std.

There would also be value in a more complex interface that supports
NUMA, Windows "processor groups", and any other mechanism that *doesn't*
work as simply as "create that many threads". That interface absolutely
needs to bake in crates.io for a while before we could consider what
subset of it makes sense in std.

SimonSapin commented:
> That seems reasonable. Another case where these might differ would be when running in a Linux cgroup that limits available CPUs.

Does the implementation in this PR return the appropriate number, when run from such a cgroup?

retep998 commented:
Another case where they might differ is on Windows when the process has a process affinity or is run inside a job object that constrains the affinity.

the8472 commented:
> 
> 
> > That seems reasonable. Another case where these might differ would be when running in a Linux cgroup that limits available CPUs.
> 
> Does the implementation in this PR return the appropriate number, when run from such a cgroup?

No, see previous comments (https://github.com/rust-lang/rust/pull/74480#issuecomment-660516686, https://github.com/rust-lang/rust/pull/74480#discussion_r456827199)

vertexclique commented:
I propose here to have a device API, in or outside of the std, this can be started as effort outside of the std and then moved into std when it is ready to be merged in. Device API can expose something similar to what [hwloc-ls](https://linux.die.net/man/1/hwloc-ls) presents and it would be nice if it is easily programmatically queryable.

the8472 commented:
I suppose a flag argument that initially only supports `0` or some other default would not be rusty enough?

bors commented:
:umbrella: The latest upstream changes (presumably #73265) made this pull request unmergeable. Please resolve the merge conflicts.

Dylan-DPC commented:
@yoshuawuyts any updates on this?

yoshuawuyts commented:
@Dylan-DPC We've chatted about the naming on Zulip and we've agreed to rename the API. I still need to push that change.

But other than that I think this patch exists more or less in it's intended form: a simple API to determine how many threads to spawn. After I've pushed the change I think it would be ready for review from the libs team.

retep998 commented:
It still needs a big warning in the documentation that the number may not be accurate. If interpreted as the number of actual hardware threads, it can undercount on Windows systems with more than 64 hardware threads. If interpreted as the available concurrency for that process, it can overcount on Windows systems when limited by a process wide affinity mask or job object limitations.

JohnCSimon commented:
Ping from triage: 
@yoshuawuyts - this PR has sat idle for the last 20 days, can you please address the comments and merge conflicts?

yoshuawuyts commented:
@JohnCSimon yep! Planning to pick this back up on Friday!

yoshuawuyts commented:
I've renamed the API to `std::thread::available_concurrency` as discussed on Zulip, included the notes on windows-specific behavior, and rebased on the latest master. If CI passes I think this should be good for review!

the8472 commented on library/std/src/thread/available_concurrency.rs:
```diff
@@ -0,0 +1,155 @@
+/// If interpreted as the number of actual hardware threads, it can undercount on
+/// Windows systems with more than 64 hardware threads. If interpreted as the
+/// available concurrency for that process, it can overcount on Windows systems
+/// when limited by a process wide affinity mask or job object limitations.
```
On linux it may also overcount due to cgroup limits or affinity.

yoshuawuyts replied:
Good point; updated the text to include that.

luser commented:
I don't want to bikeshed more as I think this is a valuable addition to `std` and you've done due diligence here, but I do think that it'd be unfortunate to ship the naive implementation on Linux that overcounts in the face of cpuset. [Docker exposes cpuset as an option to `docker run`](https://docs.docker.com/engine/reference/run/#cpuset-constraint), that seems like a very common use case:

```
$ getconf _NPROCESSORS_ONLN
8
$ docker run -ti --rm ubuntu:19.04 getconf _NPROCESSORS_ONLN
8
$ docker run -ti --rm --cpuset-cpus="0-2" ubuntu:19.04 getconf _NPROCESSORS_ONLN
8
$ docker run -ti --rm --cpuset-cpus="0-2" ubuntu:19.04 /bin/sh -c 'taskset -c -p $$'
pid 1's current affinity list: 0-2
```

That being said, this is an implementation detail and could be changed after the fact, so I think it would be reasonable to land this and file a followup issue to make the implementation more correct.

yoshuawuyts commented:
> That being said, this is an implementation detail and could be changed after the fact, so I think it would be reasonable to land this and file a followup issue to make the implementation more correct.

That's exactly how I was hoping this would go! I've created a heading in the tracking issue to track known platform limitations: https://github.com/rust-lang/rust/issues/74479.

crlf0710 commented:
Triage: It seems this is ready for review.

dtolnay reviewed (approved):
This looks reasonable to land unstable. Thanks @yoshuawuyts. The implementation here is just moved from libtest but we can follow up in subsequent PRs with behavior improvements or platform support, including the points raised by @luser and others which are already captured in the tracking issue.

dtolnay commented:
@bors r+

bors commented:
:pushpin: Commit d9207d87b346011959ff80cf86dacf2b741c61c5 has been approved by `dtolnay`

<!-- @bors r=dtolnay d9207d87b346011959ff80cf86dacf2b741c61c5 -->
<!-- homu: {"type":"Approved","sha":"d9207d87b346011959ff80cf86dacf2b741c61c5","approver":"dtolnay"} -->

JohnTitor commented:
Failed in https://github.com/rust-lang-ci/rust/runs/1261876508: 
```
 error: expected `;`, found keyword `if`
   --> library/std/src/thread/available_concurrency.rs:112:18
    |
112 |                 }
    |                  ^ help: add `;` here
...
115 |                 if res == -1 {
    |                 -- unexpected token

error[E0308]: mismatched types
   --> library/std/src/thread/available_concurrency.rs:115:27
    |
115 |                 if res == -1 {
    |                           ^^ expected `()`, found integer

error[E0308]: mismatched types
   --> library/std/src/thread/available_concurrency.rs:121:53
    |
121 |             Ok(unsafe { NonZeroUsize::new_unchecked(cpus) })
    |                                                     ^^^^ expected `usize`, found `u32`
    |
help: you can convert an `u32` to `usize` and panic if the converted value wouldn't fit
    |
121 |             Ok(unsafe { NonZeroUsize::new_unchecked(cpus.try_into().unwrap()) })
    |                                                     ^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0308`.
```

@bors r-

yoshuawuyts commented:
Ah, it seems the FreeBSD code block contained errors. I don't have a local setup to test this on, but I _think_ this should be good to try again. I believe the command to do that is (fingers crossed):

@bors try

bors commented:
@yoshuawuyts: :key: Insufficient privileges: not in try users

bors commented:
@yoshuawuyts: :key: Insufficient privileges: not in try users

JohnTitor commented:
IIRC we don't have FreeBSD builder on the try build so let's just re-add this to the queue :)
@bors r=dtolnay

bors commented:
:pushpin: Commit 9508c321c66a26711f46ab795e26f4bef0bfb8c2 has been approved by `dtolnay`

<!-- @bors r=dtolnay 9508c321c66a26711f46ab795e26f4bef0bfb8c2 -->
<!-- homu: {"type":"Approved","sha":"9508c321c66a26711f46ab795e26f4bef0bfb8c2","approver":"dtolnay"} -->

ehuss commented:
@bors r-

This will still fail:

```
error[E0308]: mismatched types
   --> library/std/src/thread/available_concurrency.rs:115:27
    |
115 |                 if res == -1 {
    |                           ^^ expected `()`, found integer

error: aborting due to previous error
```

The semicolon in the unsafe block makes the `res` variable equal to `()`.

If you have Docker, I would encourage trying to test using that.

RalfJung commented:
Alternative you can use Miri:
`XARGO_RUST_SRC=~/src/path/to/rust/library cargo miri setup --target xxx`
This does a check-only build so no linker or build tools for the target are required.

yoshuawuyts commented:
> The semicolon in the unsafe block makes the res variable equal to ().

@ehuss Ah yeah, good catch. I thought that followed out of an earlier error, but I got it wrong. That should be fixed now. Apparently I'd also missed that for OpenBSD which is fixed now too.

> If you have Docker, I would encourage trying to test using that.

Do we have instructions on how to run this available? The Windows instructions for the compiler ended up being based on my notes, and researching that took a fair bit of work. If we don't have instructions on how to compile locally on BSD on Windows yet we should probably open up an issue for this somewhere (not sure if `rust-lang/rust` is the right repo for that?)

ehuss commented:
> Do we have instructions on how to run this available?

https://rustc-dev-guide.rust-lang.org/tests/intro.html#testing-with-docker-images

The command for freebsd is `src/ci/docker/run.sh dist-x86_64-freebsd`.  Unfortunately there is no setup for openbsd.  I have an openbsd VM I occasionally use for testing, but I have never built rustc on it, and I suspect it is not easy.  Using miri also sounds useful, though I don't have much experience with that.

yoshuawuyts commented:
@RalfJung that seemed to work successfully for both FreeBSD and OpenBSD! Though from the output it seems like it may only have checked `core` but not `std`? For posterity and anyone trying this on Windows, the way I did was using the following:

```powershell
PS C:\Users\yoshu\Code\rust> rustup override set nightly-x86_64-pc-windows-gnu # miri doesn't seem to work yet on msvc
PS C:\Users\yoshu\Code\rust> rustup component add miri
PS C:\Users\yoshu\Code\rust> $env:XARGO_RUST_SRC='C:\Users\yoshu\Code\rust\library'
PS C:\Users\yoshu\Code\rust> cargo miri setup --target x86_64-unknown-freebsd
PS C:\Users\yoshu\Code\rust> cargo miri setup --target x86_64-unknown-openbsd
```

---

@ehuss dang, that seems to assume a Linux environment and I'm running on Windows. I don't think I'll be able to get that to work without provisioning a new environment (VM or otherwise) which I don't have the bandwidth for right now.

Given all known issues have been addressed at this point, can we give this a shot at running through bors again?

RalfJung commented:
> that seemed to work successfully for both FreeBSD and OpenBSD! Though from the output it seems like it may only have checked core but not std?

It should check core, std, and even test. And the output looks like that for me:
```
$ cargo miri setup --target x86_64-unknown-freebsd
   Compiling compiler_builtins v0.1.35
    Checking core v0.0.0 (/home/r/.rustup/toolchains/miri/lib/rustlib/src/rust/library/core)
   Compiling libc v0.2.79
   Compiling cc v1.0.60
   Compiling std v0.0.0 (/home/r/.rustup/toolchains/miri/lib/rustlib/src/rust/library/std)
   Compiling unwind v0.0.0 (/home/r/.rustup/toolchains/miri/lib/rustlib/src/rust/library/unwind)
    Checking rustc-std-workspace-core v1.99.0 (/home/r/.rustup/toolchains/miri/lib/rustlib/src/rust/library/rustc-std-workspace-core)
    Checking cfg-if v0.1.10
    Checking alloc v0.0.0 (/home/r/.rustup/toolchains/miri/lib/rustlib/src/rust/library/alloc)
    Checking rustc-demangle v0.1.16
    Checking panic_abort v0.0.0 (/home/r/.rustup/toolchains/miri/lib/rustlib/src/rust/library/panic_abort)
    Checking rustc-std-workspace-alloc v1.99.0 (/home/r/.rustup/toolchains/miri/lib/rustlib/src/rust/library/rustc-std-workspace-alloc)
    Checking panic_unwind v0.0.0 (/home/r/.rustup/toolchains/miri/lib/rustlib/src/rust/library/panic_unwind)
    Checking hashbrown v0.9.0
    Finished release [optimized] target(s) in 25.33s
    Checking rustc-std-workspace-std v1.99.0 (/home/r/.rustup/toolchains/miri/lib/rustlib/src/rust/library/rustc-std-workspace-std)
    Checking proc_macro v0.0.0 (/home/r/.rustup/toolchains/miri/lib/rustlib/src/rust/library/proc_macro)
    Checking term v0.0.0 (/home/r/.rustup/toolchains/miri/lib/rustlib/src/rust/library/term)
    Checking unicode-width v0.1.8
    Checking getopts v0.2.21
    Checking test v0.0.0 (/home/r/.rustup/toolchains/miri/lib/rustlib/src/rust/library/test)
    Finished release [optimized] target(s) in 1.79s
```

ehuss commented:
@bors r=dtolnay

bors commented:
:pushpin: Commit 3717646366a91ee9ca0b574973ec090670f9a4f8 has been approved by `dtolnay`

<!-- @bors r=dtolnay 3717646366a91ee9ca0b574973ec090670f9a4f8 -->
<!-- homu: {"type":"Approved","sha":"3717646366a91ee9ca0b574973ec090670f9a4f8","approver":"dtolnay"} -->

bors commented:
:hourglass: Testing commit 3717646366a91ee9ca0b574973ec090670f9a4f8 with merge c38ddb8040edce1b05bc09a0e8439472e9f67623...
<!-- homu: {"type":"BuildStarted","head_sha":"3717646366a91ee9ca0b574973ec090670f9a4f8","merge_sha":"c38ddb8040edce1b05bc09a0e8439472e9f67623"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1270238776), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=37289)
Approved by: dtolnay
Pushing c38ddb8040edce1b05bc09a0e8439472e9f67623 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"dtolnay","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=37289","checks-actions":"https://github.com/rust-lang-ci/rust/runs/1270238776"},"merge_sha":"c38ddb8040edce1b05bc09a0e8439472e9f67623"} -->
